### DISCUSSION TITLE
creating a visual scripting editor

### DISCUSSION THREAD

**Discussion by quantumgames-studiosinc** [playcanvas/engine#4094] [Show and tell]
i am creating a visual scripting editor for this. devs, i would love to see this incorporated to the engine once it is done, so if you could reply, that would be great! github link for information and updates: https://github.com/quantumgames-studiosinc/flowcanvas-playcanvas-visual-scripting-system

**Comment by yaustar** [comment]
You are more than welcome to make a PR to the engine for it when your visual editor is ready. Having a quick look, you may be able to make the integration a separate library / file such as how 8th Wall and Zappar have done theirs.

**Comment by kungfooman** [comment]
Just some random thoughts:

1) Why do you commit node_modules?
2) Why do you wrap everything in:

```js
(function(global) {
    var LiteGraph = global.LiteGraph;
    // ...
})(this);
```
3) Decide on tabs or spaces, currently its mixed, so reading code on GitHub is skewed
4) ES6 runs in every modern browser and you don't need annoying build tools in the first place. You can still use bundlers and minifiers, just please don't make them necessary.
5) One file = one class, you squeeze many things into one file
6) JSDoc typing is quite advanced and helps you to be more efficient 
7) Immediate feedback is valuable, like a transparant editor over the entire game screen (easy to toggle), see my YouTube video
8) As useful vec2/vec3's are, usually the more interesting stuff that needs debugging is matrices/quats
9) I like to see the values flowing through the actual nodes, like displaying 4x4 matrices and accessing them all individually
10) You can make use of newer JS features to get rid of unnecessary long code examples like:

```js
  var A = this.getInputData(0);
  if( A === undefined )
    A = 0;
  var B = this.getInputData(1);
  if( B === undefined )
    B = 0;
```

Which is e.g. the same as:

```js
const A = this.getInputData(0) ?? 0;
const B = this.getInputData(1) ?? 0;
```

I worked out something similiar some time ago, maybe you get some UI ideas: 
https://youtu.be/NymucpJ7V18
